src/Control/Carrier/TaskPool.hs

```diff
@@ -69,25 +76,31 @@ stopWhenDone State{..} act = do
         else act
     _ -> act
 
-worker :: (MonadIO m, Has (Lift IO) sig m) => State m -> m ()
+worker :: (Has (Lift IO) sig m) => State m -> m ()
 worker st@State{..} = loop
   where
 
-  loop = join $ liftIO $ atomically $ stopWhenDone st $ do
+  loop = join $ sendIO $ atomically $ stopWhenDone st $ do
     queued <- readTVar stQueued
     case queued of
       [] -> retry
       (x:xs) -> do
         writeTVar stQueued xs
         addRunning
-        -- FIXME: async exceptions should kill execution
-        pure $ x `finally` (complete *> loop)
+        pure $
+          mask $ \restore -> do
```

Comment:
`mask` is confusing, but it essentially masks (ignores) exceptions in the inner block, which is useful for cleanup actions like filling the mutable variable. `restore` is used to "unmask"  exceptions for the actual task